daemon/api_themes.go

```diff
@@ -0,0 +1,245 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package daemon
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/store"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var (
+	themesCmd = &Command{
+		Path:   "/v2/themes",
+		UserOK: true,
+		GET:    checkThemes,
+		POST:   installThemes,
+	}
+)
+
+type themeStatus string
+
+const (
+	themeInstalled   themeStatus = "installed"
+	themeAvailable   themeStatus = "available"
+	themeUnavailable themeStatus = "unavailable"
+)
+
+type themeStatusResponse struct {
+	GtkThemes   map[string]themeStatus `json:"gtk-themes"`
+	IconThemes  map[string]themeStatus `json:"icon-themes"`
+	SoundThemes map[string]themeStatus `json:"sound-themes"`
+}
+
+func getInstalledThemes(d *Daemon) (gtkThemes, iconThemes, soundThemes []string, err error) {
+	infos := d.overlord.InterfaceManager().Repository().Info(&interfaces.InfoOptions{
+		Names: []string{"content"},
+		Slots: true,
+	})
+	for _, info := range infos {
+		for _, slot := range info.Slots {
+			var content string
+			// The content interface ensures this attribute exists
+			if err := slot.Attr("content", &content); err != nil {
+				return nil, nil, nil, err
+			}
+			var themes *[]string
+			switch content {
+			case "gtk-3-themes":
```

Comment:
This has `gtk-3-themes` but a search below is for snap named `gtk-themes-*`. Is it too late to fix this discrepancy?

Reply:
The `gtk-3-themes` name is locked in by the many application snaps.  As mentioned earlier, due to the way theme authors maintain their themes there isn't much reason to pick a package naming scheme that would double the maintenance work.  Not to mention what will happen if themes need new versions for GTK 4.

Reply:
what will indeed happen if there's a GTK4 , the themes have a gtk-3-themes content but they are called just gtk-theme ? will they grow to contain support for both? so that install gtk-theme-foo will give content slots for both gtk-3-themes and gtk-4-themes,  or will we need to retrofit ways to find gtk3-only vs both vs gtk4-only themes?

Reply:
I don't expect GTK 4 to be any different: there won't be separate settings to select a GTK 4 theme independent of a GTK 3 theme, and if there is a matching GTK 4 theme it will most likely be maintained by the same authors as the other versions.